NEP: propose new nep for allocator policies (#18805)
numpy/numpy: doc/neps/nep-0049.rst

own. Two such use-cases are to ensure data alignment and to pin certain
allocations to certain NUMA cores.

User who wish to change the NumPy data memory management routines will use
:c:func:`PyDataMem_SetHandler`, which uses a :c:type:`PyDataMem_Handler`
structure to hold pointers to functions used to manage the data memory. The
calls are wrapped by internal routines to call :c:func:`PyTraceMalloc_Track`,
:c:func:`PyTraceMalloc_Untrack`, and will use the
:c:func:`PyDataMem_EventHookFunc` mechanism  already present in NumPy for
auditing purposes. 

Since a call to ``PyDataMem_SetHandler`` will change the default functions, but
that function may be called during the lifetime of an ``ndarray`` object, each
``ndarray`` will carry with it the ``PyDataMem_Handler`` struct used at the
time of its instantiation, and these will be used to reallocate or free the
data memory of the instance. Internally NumPy may use ``memcpy` or ``memset``
on the data ``ptr``.

Usage and Impact
----------------

The new functions can only be accessed via the NumPy C-API. An example is
included later in the NEP. The added ``struct`` will increase the size of the

> ``ndarray`` object. It is one of the major drawbacks of this approach. We can

Reviewer: Why call it a "major drawback" if it's backwards compatible? Unless there's serious impact on, e.g., use cases for small arrays where memory usage or performance would be degraded to a noticeable extent, I don't think "major drawback" is appropriate. And if it is, you should explain the actual impact.

Author: "major drawback" -> "necessary price"

    Sets the allocation event hook for numpy array data.
  
    Returns a pointer to the previous hook or NULL.  If old_data is
    non-NULL, the previous user_data pointer will be copied to it.
  
    If not NULL, hook will be called at the end of each PyDataMem_NEW/FREE/RENEW:

    .. code-block:: c
   
        result = PyDataMem_NEW(size)        -> (*hook)(NULL, result, size, user_data)
        PyDataMem_FREE(ptr)                 -> (*hook)(ptr, NULL, 0, user_data)
        result = PyDataMem_RENEW(ptr, size) -> (*hook)(ptr, result, size, user_data)
  
    When the hook is called, the GIL will be held by the calling
    thread.  The hook should be written to be reentrant, if it performs
    operations that might cause new allocation events (such as the
    creation/destruction numpy objects, or creating/destroying Python
    objects which might cause a gc)

Sample code
===========

This code adds a 64-byte header to each ``data`` pointer and stores information
about the allocation in the header. Before calling ``free``, a check ensures
the ``sz`` argument is correct ::

> 

Reviewer: This could use `.. code-block:: c`

Author: ok

    { NULL }
    };

    static struct PyModuleDef moduledef = {
        PyModuleDef_HEAD_INIT,
        "mem_policy",  /* m_name */
        NULL,           /* m_doc */
        -1,             /* m_size */
        methods,        /* m_methods */
    };

    PyMODINIT_FUNC
    PyInit_mem_policy(void) {
    PyObject *mod = PyModule_Create(&moduledef);
        import_array();
        return mod;
    }
 

Related Work
------------

The NEP is being tracked by the pnumpy_ project and a `comment in the PR`_
mentions use in orchestrating FPGA DMAs.

Implementation
--------------

The NEP has been implemented in `PR  17582`_.

Alternatives
------------

These were discussed in `issue 17467`_. `PR 5457`_ proposed a
global interface for specifying aligned allocations. Alignment can be
crucial for some applications, but in general is just extra overhead, so it
should be configurable by the user/app. ``PyArray_malloc_aligned`` and
friends were added to NumPy with the random API refactor. and are used there
for performance.

`PR 5470`_ had two parts: configurable `PyDataMem*` overrides and a hook
mechanism. The hook mechanism was merged but the part described by this NEP

> was not.

Reviewer: Can you link to where it got merged? That complete PR is open and didn't get any review comments.

Author: Sorry, the memory hooks were merged in gh-309. I will restate the history.

===================================
NEP 49 — Data allocation strategies
===================================

:Author: Matti Picus
:Status: Draft
:Type: Standards Track
:Created: 2021-04-18

Abstract
--------

The ``numpy.ndarray`` requires additional memory allocations
to hold ``numpy.ndarray.strides``, ``numpy.ndarray.shape`` and
``numpy.ndarray.data`` attributes. These attributes are specially allocated
after creating the python object in ``__new__`` method.

This NEP proposes a mechanism to override the memory management strategy used
for ``ndarray->data`` with user-provided alternatives. This allocation holds

> the arrays data and is can be very large. As accessing this data often becomes

Reviewer 2: ```suggestion
the arrays data and can be very large. As accessing this data often becomes
```

Author: fixed

a performance bottleneck, custom allocation strategies to guarantee data
alignment or pinning allocations to specialized memory hardware can enable

> hardware-specific optimizations. The other allocatations remain unchanged.

Reviewer 2: ```suggestion
hardware-specific optimizations. The other allocations remain unchanged.
```

Author: fixed

Motivation and Scope
--------------------

Users may wish to override the internal data memory routines with ones of their
own. Two such use-cases are to ensure data alignment and to pin certain

> allocations to certain NUMA cores. This desire for alignment was disucssed

Reviewer 2: ```suggestion
allocations to certain NUMA cores. This desire for alignment was discussed
```

Author: fixed

``ndarray`` will carry with it the ``PyDataMem_Handler`` struct used at the
time of its instantiation, and these will be used to reallocate or free the
data memory of the instance. Internally NumPy may use ``memcpy`` or ``memset``
on the pointer to the data memory.

NumPy C-API functions
=====================

.. c:type:: PyDataMem_Handler

    A struct to hold function pointers used to manipulate memory

    .. code-block:: c

        typedef struct {
            char name[128];  /* multiple of 64 to keep the struct unaligned */
            PyDataMem_AllocFunc *alloc;
            PyDataMem_ZeroedAllocFunc *zeroed_alloc;
            PyDataMem_FreeFunc *free;
            PyDataMem_ReallocFunc *realloc;
        } PyDataMem_Handler;

    where the function's signatures are

    .. code-block:: c

        typedef void *(PyDataMem_AllocFunc)(size_t size);
        typedef void *(PyDataMem_ZeroedAllocFunc)(size_t nelems, size_t elsize);
        typedef void (PyDataMem_FreeFunc)(void *ptr, size_t size);
        typedef void *(PyDataMem_ReallocFunc)(void *ptr, size_t size);

.. c:function:: const PyDataMem_Handler * PyDataMem_SetHandler(PyDataMem_Handler *handler)

>    Sets a new allocation policy. If the input value is NULL, will reset
>    the policy to the default. Returns the previous policy, NULL if the
>    previous policy was the default. We wrap the user-provided functions
>    so they will still call the Python and NumPy memory management callback
>    hooks. All the function pointers must be filled in, NULL is not accepted.

Reviewer 2: ```suggestion
   Sets a new allocation policy. If the input value is ``NULL``, will reset
   the policy to the default. Returns the previous policy, ``NULL`` if the
   previous policy was the default. We wrap the user-provided functions
   so they will still call the Python and NumPy memory management callback
   hooks. All the function pointers must be filled in, ``NULL`` is not accepted.
```

Author: fixed

to the ``ndarray->data`` pointer were already breaking the current memory
management strategy and should restore
``ndarray->data`` before calling ``Py_DECREF``. As mentioned above, the change
in size should not impact end-users.

Detailed description
--------------------

High level design
=================

Users who wish to change the NumPy data memory management routines will use
:c:func:`PyDataMem_SetHandler`, which uses a :c:type:`PyDataMem_Handler`
structure to hold pointers to functions used to manage the data memory.

Since a call to ``PyDataMem_SetHandler`` will change the default functions, but
that function may be called during the lifetime of an ``ndarray`` object, each
``ndarray`` will carry with it the ``PyDataMem_Handler`` struct used at the
time of its instantiation, and these will be used to reallocate or free the
data memory of the instance. Internally NumPy may use ``memcpy`` or ``memset``
on the pointer to the data memory.

NumPy C-API functions
=====================

.. c:type:: PyDataMem_Handler

    A struct to hold function pointers used to manipulate memory

    .. code-block:: c

        typedef struct {

>             char name[128];  /* multiple of 64 to keep the struct aligned */

Reviewer 3: I wasn't aware this had any effect - doesn't an array in a strict only contribute it's element size to the alignment?

Reviewer 4: @eric-wieser I understood it as relying on the assumption that the struct itself is 64 byte (cache-line) aligned. In that case the interesting part of the struct will remain aligned if the `name` is a multiple of 64 byte.

But, I might be misunderstanding it and I think you are right that it may make sense to add some attributes/pragma to ensure the struct itself is cache-line aligned.  (Although, that is not part of the header.)

Reviewer 3: On another note - Python doesn't use static char buffers for any of its object names, as I remember. Is there a reason for doing so here rather than using a `char*`?

Author: How can we make it clear who allocates and frees the content of the `char *` and prevent corruption?

Reviewer 3: I just posted to the mailing list with a related comment, which I'll repeat here for ease:

> Consider an allocator that aligns to `N` bytes, where `N` is configurable from a python call in someone else's extension module. Where do they store `N`?
> They can hide it in `PyDataMem_Handler::name` but that's obviously an abuse of the API.
> They can store it as a global variable, but then obviously the idea of tracking the allocator used to construct an array doesn't work, as the state ends up changing with the global allocator.
> 
> The easy way out here would be to add a `void* context` field to the structure, and pass it into all the methods.
> This doesn't really solve the problem though, as now there's no way to cleanup any allocations used to populate `context`, or worse decrement references to python objects stored within `context`.
> I think we want to bundle `PyDataMem_Handler` in a `PyObject` somehow, either via a new C type, or by using the PyCapsule API which has the cleanup and state hooks we need.
> `PyDataMem_GetHandlerName` would then return this PyObject rather than an opaque name.

Reviewer 3: So I guess more concretely, my suggestion is:
```python
typedef void *(PyDataMem_AllocFunc)(PyObject *self, size_t size);
typedef void *(PyDataMem_ZeroedAllocFunc)(PyObject *self, size_t nelems, size_t elsize);
typedef void (PyDataMem_FreeFunc)(PyObject *self, void *ptr, size_t size);
typedef void *(PyDataMem_ReallocFunc)(PyObject *self, void *ptr, size_t size);

typedef struct {
    PyObject_HEAD
    PyDataMem_AllocFunc *alloc;
    PyDataMem_ZeroedAllocFunc *zeroed_alloc;
    PyDataMem_FreeFunc *free;
    PyDataMem_ReallocFunc *realloc;
} PyDataMem_HandlerObject;

// I haven't used the CAPI in a while, but there's not much boilerplate here anyway
PyType_Spec PyDataMem_HandlerTypeSpec =
{
    .name = "numpy.core.datamemhandler",
    .basicsize = sizeof(PyDataMem_HandlerObject),
    .itemsize = 0,
    .flags = Py_TPFLAGS_BASETYPE
};

/-- ALlocate a 
PyDataMem_HandlerObject* PyDataMem_Handler_New() {
   // whatever the boilerplate API is.
}
```

Users who are fine with a stateless allocator would use:
```C
PyDataMem_HandlerObject *handler = PyObject_New(PyDataMem_HandlerObject, &PyDataMem_HandlerType)
```
and then can set the fields as appropriate.

Users who want a stateful allocator can then just create a derived type with some extra state.

Author: Responded on the mailing list.

about the allocation in the header. Before calling ``free``, a check ensures
the ``sz`` argument is correct.

.. code-block:: c

    #define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
    #include <numpy/arrayobject.h>
    NPY_NO_EXPORT void *

    shift_alloc(size_t sz) {
        char *real = (char *)malloc(sz + 64);
        if (real == NULL) {
            return NULL;
        }
        snprintf(real, 64, "originally allocated %ld", (unsigned long)sz);
        return (void *)(real + 64);
    }

    NPY_NO_EXPORT void *
    shift_zero(size_t sz, size_t cnt) {
        char *real = (char *)calloc(sz + 64, cnt);
        if (real == NULL) {
            return NULL;
        }
        snprintf(real, 64, "originally allocated %ld via zero",
                 (unsigned long)sz);
        return (void *)(real + 64);
    }

    NPY_NO_EXPORT void
    shift_free(void * p, npy_uintp sz) {
        if (p == NULL) {
            return ;
        }
        char *real = (char *)p - 64;
        if (strncmp(real, "originally allocated", 20) != 0) {
            fprintf(stdout, "uh-oh, unmatched shift_free, "
                    "no appropriate prefix\\n");

>             /* Make gcc crash by calling free on the wrong address */

Reviewer 3: I assume by gcc you mean "the c runtime"?

Author: fixed

===================================
NEP 49 — Data allocation strategies
===================================

:Author: Matti Picus
:Status: Draft
:Type: Standards Track
:Created: 2021-04-18

Abstract
--------

The ``numpy.ndarray`` requires additional memory allocations
to hold ``numpy.ndarray.strides``, ``numpy.ndarray.shape`` and
``numpy.ndarray.data`` attributes. These attributes are specially allocated
after creating the python object in ``__new__`` method. The ``strides`` and
``shape`` are stored in a piece of memory allocated internally.

These allocations are small relative to the ``data``, the homogeneous chunk of
memory used to store the actual array values (which could be pointers in the
case of ``object`` arrays). This NEP proposes a mechanism to override the
default memory management strategy in NumPy with user-provided alternatives.

Motivation and Scope
--------------------

Users may wish to override the internal data memory routines with ones of their
own. Two such use-cases are to ensure data alignment and to pin certain
allocations to certain NUMA cores. The NEP is based on `PR 17582`_ and

> largely taken from the documentation added there.

Author: ok

User who wish to change the NumPy data memory management routines will use
:c:func:`PyDataMem_SetHandler`, which uses a :c:type:`PyDataMem_Handler`
structure to hold pointers to functions used to manage the data memory. The
calls are wrapped by internal routines to call :c:func:`PyTraceMalloc_Track`,
:c:func:`PyTraceMalloc_Untrack`, and will use the

> :c:func:`PyDataMem_EventHookFunc` mechanism  already present in NumPy for

Author: this was removed, are you looking at an older version?

> :c:func:`PyDataMem_EventHookFunc` mechanism  already present in NumPy for

Reviewer 2: Sorry, this may have been from an unfinished prior review.

auditing purposes. 

Since a call to ``PyDataMem_SetHandler`` will change the default functions, but
that function may be called during the lifetime of an ``ndarray`` object, each
``ndarray`` will carry with it the ``PyDataMem_Handler`` struct used at the
time of its instantiation, and these will be used to reallocate or free the
data memory of the instance.

Usage and Impact
----------------

The new functions can only be accessed via the NumPy C-API. An example is
included later in the NEP. The added ``struct`` will increase the size of the
``ndarray`` object. It is one of the major drawbacks of this approach. We can
be reasonably sure that the change in size will have a minimal impact on

> end-user code because NumPy version 1.20 already changed the object size.

Reviewer 4: Seems fine here, in line with my questions below, it might be nice to have one paragraph that explains how users will benefit (that is probably the users of the libraries that modify the allocation strategy).

I assume this are things like: Alignment guarantees or NUMA core pinning can speed up their code. Allocating in memory shared by GPU and CPU speeds up certain operations?

Author: > Alignment guarantees or NUMA core pinning can speed up their code

Yes. There are libraries like `pnumpy` that would like to keep data on a specific core or hardware where unaligned memory is expensive (like ARMv7).

Author: I think using pinned memory would hit some snags due to `memcpy`. A future enhancement might be to enable user-supplied `memcpy` routines and using them inside numPy when creating a copy of an ndarray.

Reviewer 5: Just FYI. CuPy recently added a few APIs to allocate pinned memory that backs a NumPy array: 
> For using pinned memory more conveniently, we also provide a few high-level APIs in the cupyx namespace, including cupyx.empty_pinned(), cupyx.empty_like_pinned(), cupyx.zeros_pinned(), and cupyx.zeros_like_pinned(). They return NumPy arrays backed by pinned memory. If CuPy’s pinned memory pool is in use, the pinned memory is allocated from the pool.

https://docs.cupy.dev/en/stable/user_guide/memory.html#memory-management. I think PyCUDA also have similar APIs. 

Indeed, memcpy could be an issue for D2H transfers, but not H2D (at least in CuPy, `arr_cp[:] = arr_np_on_pinned` would handle this properly). It's something I am working to circumvent (by calling `arr_np_on_pinned = cp.asnumpy(arr_cp, out=arr_np_on_pinned)`).

Backward compatibility
----------------------

The design will not break backward compatibility. Projects that were assigning
to the ``ndarray->data`` pointer were already breaking the current memory
management strategy (backed by ``npy_alloc_cache``) and should restore
``ndarray->data`` before calling ``Py_DECREF``. As mentioned above, the change

> in size should not impact end-users.

Reviewer 4: This is not quite true. The old strategy was backward compatible. The worst thing that could happen is that you allocate a huge chunk, create an array with it, set OWNDATA and then the free will put it into a "too large" bucket (wasting memory).  But overall, there was never any API change the way the cache was implemented.

This PR has a tiny chance of that, because _if_ an external project would replace `ndarray->data` (i.e. realloc or use free/malloc), that would have worked before but it will stop working if the allocation strategy was modified.

Now, I am not worried about that. Nobody should be doing that, and it can only be a problem if you use one project modifying the allocation strategy AND a project modifying `ndarray->data` together.

Your PR makes sure that if the user provides the data (and may have allocated it), we will end up using the normal `free` strategy, so no issue exists for arrays created from passed in `data`.

Reviewer 4: (same also applies to `->shape`, but that is even crazier, since the shape and strides were always allocated together, so you have to dig into NumPy code to get it right.  Overall, it makes me almost wonder if we could get away with breaking ABI and replacing the allocation strategy there with `PyMem_Malloc()` by default, but I guess it doesn't matter.)

Author: I didn't change the shape/strides allocation in the PR. Should I make it explicit in the NEP that the policy only touches the data and not the shape/strides ?

Reviewer 4: Yeah, I think I misread the abstract a bit and got sidetracked.  Maybe rearrange the second paragraph of the abstract slightly would set the track clearer (brainstorming here):

> This NEP proposes a mechanism to override the memory management strategy used for `ndarray->data` with user-provided alternatives.  This allocation holds the arrays data and is can be very large.  As this is the data that operations work on it can benefit from custom allocation strategies to guarantee data alignment or placement in memory shared with the GPU.

Author: Updated.

Detailed description
--------------------

NumPy C-API functions
=====================

As of NumPy version 1.21, the copy functions are not yet implemented, all
memory copies are handled by calls to ``memcpy``. The API leaves room for a
future enhancement which will carefully check all the ``memcpy`` calls in NumPy

> and replace them with ``host2obj``, ``obj2obj``, or ``obj2host``.

Reviewer 4: I am still a bit confused by these.  I agree with leaving room at the end, but I wonder if we should name them already.  And if we should leave _more_ room in case we have another idea?  (I guess, I would like to understand them better if they are so clearly prepared, but I currently don't; but if others are happy also fine.)

I assume it is expected that users may leave them `NULL` at this point as opposed to setting them to `memcpy` in anticipation that they may be used at some point?  Why not just call it `void *reserved[3]` at the end and explicitly allow leaving them `NULL`? (which I guess may mean that they can be filled in during registration!

Author: I will remove them, and document that internally `memcpy` or `memset` may be called on the `data` pointer.

.. c:type:: PyDataMem_Handler

    A struct to hold function pointers used to manipulate memory

    .. code-block:: c

        typedef struct {

>             char name[128];  /* multiple of 64 to keep the struct unaligned */

Author: fixed

         return PyUnicode_FromString(old->name);
     
    };
        
    static PyMethodDef methods[] = {
    {"test_prefix", (PyCFunction)mem_policy_test_prefix, METH_O},
    {"set_new_policy", (PyCFunction)mem_policy_set_new_policy, METH_NOARGS},
    {"set_old_policy", (PyCFunction)mem_policy_set_old_policy, METH_NOARGS},
    { NULL }
    };

    static struct PyModuleDef moduledef = {
        PyModuleDef_HEAD_INIT,
        "mem_policy",  /* m_name */
        NULL,           /* m_doc */
        -1,             /* m_size */
        methods,        /* m_methods */
    };

    PyMODINIT_FUNC
    PyInit_mem_policy(void) {
    PyObject *mod = PyModule_Create(&moduledef);
        import_array();
        return mod;
    }
 

Related Work
------------

The NEP is being tracked by the pnumpy_ project and a `comment in the PR`_
mentions use in orchestrating FPGA DMAs.

Implementation
--------------

The NEP has been implemented in `PR  17582`_.

Alternatives
------------

These were discussed in `issue 17467`_. `PR 5457`_ proposed a
global interface for specifying aligned allocations. But alignment can be
crucial for some applications, but in general is just extra overhead, so it

> should be configurable by the user/app.

Reviewer 4: Hmmm, not sure I follow.  Is this that you want to distinguish between small/large arrays but that would be difficult or impossible with a global "please align to XY" setting?

Author: Yes

`PR 5470`_ had two parts: configurable `PyDataMem*` overrides and a hook
mechanism. The hook mechanism was merged (does anyone use that?) but the part I
am interested in would requre an API change.

A fallback proposal would be to allow specifying a default allocation alignment
aka ``PR 5457``, perhaps using ``PyArray_malloc_aligned`` and friends which

> were added to NumPy with the random API refactor.

Reviewer 4: This seems to duplicate the first paragraph, can probably just remove it.

===================================
NEP 49 — Data allocation strategies
===================================

:Author: Matti Picus
:Status: Draft
:Type: Standards Track
:Created: 2021-04-18

Abstract
--------

The ``numpy.ndarray`` requires additional memory allocations
to hold ``numpy.ndarray.strides``, ``numpy.ndarray.shape`` and
``numpy.ndarray.data`` attributes. These attributes are specially allocated
after creating the python object in ``__new__`` method. The ``strides`` and

> ``shape`` are stored in a piece of memory allocated internally.

Reviewer 5: As per https://github.com/numpy/numpy/pull/18805#discussion_r616230941, can we rephrase/remove this paragraph please? I misread this too and thought the new allocator interface would change how `shape` and `strides` are allocated, which I couldn't care less.

This NEP proposes a mechanism to override the memory management strategy used
for ``ndarray->data`` with user-provided alternatives. This allocation holds
the arrays data and is can be very large. As accessing this data often becomes
a performance bottleneck, custom allocation strategies to guarantee data
alignment or pinning allocations to specialized memory hardware can enable
hardware-specific optimizations.

Motivation and Scope
--------------------

Users may wish to override the internal data memory routines with ones of their
own. Two such use-cases are to ensure data alignment and to pin certain
allocations to certain NUMA cores. This desire for alignment was disucssed
multiple times on the mailing list `in 2005`_,  and in `issue 5312`_ in 2014,

> which led to `PR 5457` and more mailing list discussions here_ `and here`_. In

Reviewer 5: nitpick: we need to link to `PR 5457`.

a comment on the issue `from 2017`_, a user described how 64-byte alignment
improved performance by 40x.

> Also related is `issue 14177` around the use of ``madvise`` and huge pages on

Reviewer 5: nitpick: we need to link to `issue 14177`

Linux.

Various tracing and profiling libraries like filprofiler_ or `electric fence`_
override ``malloc``.

The long CPython discussion of `BPO 18835`_  began with discussing the need for
``PyMem_Alloc32`` and ``PyMem_Alloc64``.  The early conclusion was that the
cost (of wasted padding) vs. the benifit of aligned memory is best left to the
user, but then evolves into a discussion of various proposals to deal with
memory allocations, including `PEP 445`_ `memory interfaces`_ to
``PyTraceMalloc_Track`` which apparently was explictly added for NumPy.

Allowing users to implement different strategies via the NumPy C-API will
enable exploration of this rich area of possible optimizations. The intention
is to create a flexible enough interface without burdening normative users.

> _`issue 5312`: https://github.com/numpy/numpy/issues/5312 

Reviewer 5: The hyperlinks are rendered incorrectly. I think this is the correct syntax:
```suggestion
.. _issue 5312: https://github.com/numpy/numpy/issues/5312 
```
(Same applies to the links below.)

_`from 2017`: https://github.com/numpy/numpy/issues/5312#issuecomment-315234656
_`in 2005`: https://numpy-discussion.scipy.narkive.com/MvmMkJcK/numpy-arrays-data-allocation-and-simd-alignement
_`here`: http://numpy-discussion.10968.n7.nabble.com/Aligned-configurable-memory-allocation-td39712.html
_`and here`: http://numpy-discussion.10968.n7.nabble.com/Numpy-s-policy-for-releasing-memory-td1533.html
_`issue 14177`: https://github.com/numpy/numpy/issues/14177
_`filprofiler`: https://github.com/pythonspeed/filprofiler/blob/master/design/allocator-overrides.md
_`electric fence`: https://github.com/boundarydevices/efence
_`memory interfaces`: https://docs.python.org/3/c-api/memory.html#customize-memory-allocators
_`BPO 18835`: https://bugs.python.org/issue18835
_`PEP 445`: https://www.python.org/dev/peps/pep-0445/

Usage and Impact
----------------

The new functions can only be accessed via the NumPy C-API. An example is

> included later in the NEP. The added ``struct`` will increase the size of the

Reviewer 5: nitpick
```suggestion
included later in this NEP. The added ``struct`` will increase the size of the
```

> ``ndarray`` object. It is a necessary price to be paid for this approach. We

Reviewer 5: Sorry if I am being silly since I am not a native speaker, but don't we just say "price to pay"?
```suggestion
``ndarray`` object. It is a necessary price to pay for this approach. We
```

can be reasonably sure that the change in size will have a minimal impact on
end-user code because NumPy version 1.20 already changed the object size.

The implementation preserves the use of ``PyTraceMalloc_Track`` to track
allocations already present in NumPy.

Backward compatibility
----------------------

The design will not break backward compatibility. Projects that were assigning
to the ``ndarray->data`` pointer were already breaking the current memory

> management strategy (backed by ``npy_alloc_cache``) and should restore

Reviewer 5: Is there an API doc or something for ``npy_alloc_cache`` that we can link to?

Author: There is more documentation in the PR, but currently `npy_alloc_cache` is undocumented. I will remove this reference.

``ndarray->data`` before calling ``Py_DECREF``. As mentioned above, the change
in size should not impact end-users.

Detailed description
--------------------

High level design
=================

Users who wish to change the NumPy data memory management routines will use
:c:func:`PyDataMem_SetHandler`, which uses a :c:type:`PyDataMem_Handler`
structure to hold pointers to functions used to manage the data memory. The
calls are wrapped by internal routines to call :c:func:`PyTraceMalloc_Track`,
:c:func:`PyTraceMalloc_Untrack`, and will use the
:c:func:`PyDataMem_EventHookFunc` mechanism  already present in NumPy for
auditing purposes. 

> Since a call to ``PyDataMem_SetHandler`` will change the default functions, but
> that function may be called during the lifetime of an ``ndarray`` object, each
> ``ndarray`` will carry with it the ``PyDataMem_Handler`` struct used at the
> time of its instantiation, and these will be used to reallocate or free the
> data memory of the instance. Internally NumPy may use ``memcpy` or ``memset``
> on the data ``ptr``.

Reviewer 5: nitpick
```suggestion
Since a call to :c:func:`PyDataMem_SetHandler` will change the default functions, but
that function may be called during the lifetime of an ``ndarray`` object, each
``ndarray`` will carry with it the :c:type:`PyDataMem_Handler` struct used at the
time of its instantiation, and these will be used to reallocate or free the
data memory of the instance. Internally NumPy may use ``memcpy` or ``memset``
on the data ``ptr``.
```

Reviewer 5: Another question: is ``ptr`` a member of some struct?

Author: I find the multiple use of links distracting. The first mention is a link, the others are not.

> is `ptr` a member of some struct?

No, it is the return value from `alloc`. I will reformat to make it more clear.

NumPy C-API functions
=====================

.. c:type:: PyDataMem_Handler

    A struct to hold function pointers used to manipulate memory

    .. code-block:: c

        typedef struct {
            char name[128];  /* multiple of 64 to keep the struct unaligned */
            PyDataMem_AllocFunc *alloc;
            PyDataMem_ZeroedAllocFunc *zeroed_alloc;
            PyDataMem_FreeFunc *free;
            PyDataMem_ReallocFunc *realloc;

>         } PyDataMem_Handler;

Reviewer 5: Question: Can `zeroed_alloc` or `realloc` be NULL? I can imagine for simplicity an application might just wanna set `alloc` and `free`. If not, we should probably say it out loud that none of these can be NULL.

Author: They cannot be NULL.

    where the function's signatures are

    .. code-block:: c

        typedef void *(PyDataMem_AllocFunc)(size_t size);
        typedef void *(PyDataMem_ZeroedAllocFunc)(size_t nelems, size_t elsize);
        typedef void (PyDataMem_FreeFunc)(void *ptr, size_t size);
        typedef void *(PyDataMem_ReallocFunc)(void *ptr, size_t size);

>         typedef void *(PyDataMem_CopyFunc)(void *dst, const void *src, size_t size);

Reviewer 5: Is `PyDataMem_CopyFunc` a leftover from the previous edit? I don't find it in the struct.

Author: removing

.. c:function:: const PyDataMem_Handler * PyDataMem_SetHandler(PyDataMem_Handler *handler)

   Sets a new allocation policy. If the input value is NULL, will reset
   the policy to the default. Returns the previous policy, NULL if the
   previous policy was the default. We wrap the user-provided functions

>    so they will still call the python and numpy memory management callback

Reviewer 5: ```suggestion
   so they will still call the Python and NumPy memory management callback
```

   hooks.
    
.. c:function:: const char * PyDataMem_GetHandlerName(PyArrayObject *obj)

>    Return the const char name of the PyDataMem_Handler used by the
>    PyArrayObject. If NULL, return the name of the current global policy that
>    will be used to allocate data for the next PyArrayObject

Reviewer 5: ```suggestion
   Return the const char name of the :c:type:`PyDataMem_Handler` used by the
   ``PyArrayObject``. If NULL, return the name of the current global policy that
   will be used to allocate data for the next ``PyArrayObject``.
```

> For an example of setting up and using the PyDataMem_Handler, see the test in
> :file:`numpy/core/tests/test_mem_policy.py`

Reviewer 5: ```suggestion
For an example of setting up and using the :c:type:`PyDataMem_Handler`, see the test in
:file:`numpy/core/tests/test_mem_policy.py`.
```

.. c:function:: void PyDataMem_EventHookFunc(void *inp, void *outp, size_t size, void *user_data);

    This function will be called on NEW,FREE,RENEW calls in data memory

>     manipulation

Reviewer 5: This is unclear to me...Do you mean `alloc`, `free`, and `realloc`?

Reviewer 5: Or do you mean `PyDataMem_NEW` etc? If so, could you be explicit and add links to them?

Author: Removing this and the next function documentation, they are not part of the NEP.

.. c:function:: PyDataMem_EventHookFunc * PyDataMem_SetEventHook(PyDataMem_EventHookFunc *newhook, void *user_data, void **old_data)

    Sets the allocation event hook for numpy array data.
  
    Returns a pointer to the previous hook or NULL.  If old_data is

>     non-NULL, the previous user_data pointer will be copied to it.

Reviewer 5: ```suggestion
    non-NULL, the previous ``user_data`` pointer will be copied to it.
```

>     If not NULL, hook will be called at the end of each PyDataMem_NEW/FREE/RENEW:

Reviewer 5: Same here for links...

    .. code-block:: c
   
        result = PyDataMem_NEW(size)        -> (*hook)(NULL, result, size, user_data)
        PyDataMem_FREE(ptr)                 -> (*hook)(ptr, NULL, 0, user_data)
        result = PyDataMem_RENEW(ptr, size) -> (*hook)(ptr, result, size, user_data)
  
    When the hook is called, the GIL will be held by the calling
    thread.  The hook should be written to be reentrant, if it performs
    operations that might cause new allocation events (such as the

>     creation/destruction numpy objects, or creating/destroying Python

Reviewer 5: ```suggestion
    creation/destruction NumPy objects, or creating/destroying Python
```

>     objects which might cause a gc)

Reviewer 5: Better be explicit:
```suggestion
    objects which might cause a garbage collection).
```

    };

    static PyObject* mem_policy_set_new_policy(PyObject *self, PyObject *args)
    {
    
         const PyDataMem_Handler *old = PyDataMem_SetHandler(&new_handler);
         return PyUnicode_FromString(old->name);
     
    };
    
    static PyObject* mem_policy_set_old_policy(PyObject *self, PyObject *args)
    {
    
         const PyDataMem_Handler *old = PyDataMem_SetHandler(NULL);
         return PyUnicode_FromString(old->name);
     
    };
        
    static PyMethodDef methods[] = {
    {"test_prefix", (PyCFunction)mem_policy_test_prefix, METH_O},
    {"set_new_policy", (PyCFunction)mem_policy_set_new_policy, METH_NOARGS},
    {"set_old_policy", (PyCFunction)mem_policy_set_old_policy, METH_NOARGS},
    { NULL }
    };

    static struct PyModuleDef moduledef = {
        PyModuleDef_HEAD_INIT,
        "mem_policy",  /* m_name */
        NULL,           /* m_doc */
        -1,             /* m_size */
        methods,        /* m_methods */
    };

    PyMODINIT_FUNC
    PyInit_mem_policy(void) {
    PyObject *mod = PyModule_Create(&moduledef);
        import_array();
        return mod;
    }
 

Related Work
------------

> The NEP is being tracked by the pnumpy_ project and a `comment in the PR`_

Reviewer 5: ```suggestion
This NEP is being tracked by the pnumpy_ project and a `comment in the PR`_
```

mentions use in orchestrating FPGA DMAs.

Implementation
--------------

> The NEP has been implemented in `PR  17582`_.

Reviewer 5: ```suggestion
This NEP has been implemented in `PR  17582`_.
```

Alternatives
------------

These were discussed in `issue 17467`_. `PR 5457`_  and `PR 5470`_ proposed a
global interface for specifying aligned allocations. 

> ``PyArray_malloc_aligned`` and friends were added to NumPy with the random API
> refactor. and are used there for performance.

Reviewer 5: Question: is there a link to `PyArray_malloc_aligned`, or is it internal?
```suggestion
``PyArray_malloc_aligned`` and friends were added to NumPy with the ``random`` API
refactor and are used there for performance.
```

Author: > Question: is there a link to PyArray_malloc_aligned, or is it internal?

It is internal
